zerver/lib/rate_limiter.py

```diff
@@ -163,6 +176,132 @@ def rate_limit_entity(cls, entity_key: str, rules: List[Tuple[int, int]],
         # Returns (ratelimited, secs_to_freedom)
         pass
 
+class TornadoInMemoryRateLimiterBackend(RateLimiterBackend):
+    # reset_times[rule][key] is the time at which the event
+    # request from the rate-limited key will be accepted.
+    reset_times = {}  # type: Dict[Tuple[int, int], Dict[str, float]]
+
+    # last_gc_time is the last time when the garbage was
+    # collected from reset_times for rule (time_window, max_count).
+    last_gc_time = {}  # type: Dict[Tuple[int, int], float]
+
+    # timestamps_blocked_until[key] contains the timestamp
+    # up to which the key has been blocked manually.
+    timestamps_blocked_until = {}  # type: Dict[str, float]
+
+    @classmethod
+    def _garbage_collect_for_rule(cls, now: float, time_window: int, max_count: int) -> None:
+        keys_to_delete = []
+        reset_times_for_rule = cls.reset_times.get((time_window, max_count), None)
+        if reset_times_for_rule is None:
+            return
+
+        for entity_key in reset_times_for_rule:
+            if reset_times_for_rule[entity_key] < now:
+                keys_to_delete.append(entity_key)
+
+        for entity_key in keys_to_delete:
+            del reset_times_for_rule[entity_key]
+
+        if len(reset_times_for_rule) == 0:
```

Comment:
I prefer the `len() == 0` approach; I'm not a fan of using python's falsiness in this way because I'd usually rather an unexpected `None` lead to a crash than it being treated as the empty list.